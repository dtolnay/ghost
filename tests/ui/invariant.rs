#![feature(generic_const_items)]
#![allow(incomplete_features)]

use ghost::phantom;

#[phantom]
struct InvariantPhantom<#[invariant] T>;

fn require_covariant<'a>(phantom: InvariantPhantom<&'static str>) -> InvariantPhantom<&'a str> {
    phantom
}

fn require_contravariant<'a>(phantom: InvariantPhantom<&'a str>) -> InvariantPhantom<&'static str> {
    phantom
}

fn main() {}
