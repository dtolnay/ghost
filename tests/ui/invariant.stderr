error: lifetime may not live long enough
 --> tests/ui/invariant.rs:7:5
  |
6 | fn require_covariant<'a>(phantom: InvariantPhantom<&'static str>) -> InvariantPhantom<&'a str> {
  |                      -- lifetime `'a` defined here
7 |     phantom
  |     ^^^^^^^ returning this value requires that `'a` must outlive `'static`
  |
  = note: requirement occurs because of the type `InvariantPhantom__Phantom<&str>`, which makes the generic argument `&str` invariant
  = note: the enum `InvariantPhantom__Phantom<T>` is invariant over the parameter `T`
  = help: see <https://doc.rust-lang.org/nomicon/subtyping.html> for more information about variance

error: lifetime may not live long enough
  --> tests/ui/invariant.rs:11:5
   |
10 | fn require_contravariant<'a>(phantom: InvariantPhantom<&'a str>) -> InvariantPhantom<&'static str> {
   |                          -- lifetime `'a` defined here
11 |     phantom
   |     ^^^^^^^ returning this value requires that `'a` must outlive `'static`
   |
   = note: requirement occurs because of the type `InvariantPhantom__Phantom<&str>`, which makes the generic argument `&str` invariant
   = note: the enum `InvariantPhantom__Phantom<T>` is invariant over the parameter `T`
   = help: see <https://doc.rust-lang.org/nomicon/subtyping.html> for more information about variance
